jfr/src/main/java/io/netty/jfr/JfrChannelHandler.java

```diff
@@ -0,0 +1,417 @@
+            channelActiveEvent.commit();
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) {
+        final ChannelInactiveEvent channelInactiveEvent = new ChannelInactiveEvent();
+        channelInactiveEvent.begin();
+        ctx.fireChannelInactive();
+        channelInactiveEvent.end();
+        if (channelInactiveEvent.shouldCommit()) {
+            channelInactiveEvent.setChannel(ctx.channel());
+        }
+        channelInactiveEvent.commit();
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        final ExceptionCaughtEvent exceptionCaughtEvent = new ExceptionCaughtEvent();
+        exceptionCaughtEvent.begin();
+        ctx.fireExceptionCaught(cause);
+        exceptionCaughtEvent.end();
+        if (exceptionCaughtEvent.shouldCommit()) {
+            exceptionCaughtEvent.setChannel(ctx.channel());
+            exceptionCaughtEvent.setCause(cause);
+            exceptionCaughtEvent.commit();
+        }
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+        final UserEventTriggeredEvent userEventTriggeredEvent = new UserEventTriggeredEvent();
+        userEventTriggeredEvent.begin();
+        userEventTriggeredEvent.setUserEvent(evt);
+        ctx.fireUserEventTriggered(evt);
+        userEventTriggeredEvent.end();
+        if (userEventTriggeredEvent.shouldCommit()) {
+            userEventTriggeredEvent.setChannel(ctx.channel());
+            userEventTriggeredEvent.commit();
+        }
+    }
+
+    @Override
+    public void bind(ChannelHandlerContext ctx, SocketAddress localAddress, ChannelPromise promise) {
+        final BindEvent bindEvent = new BindEvent();
+        bindEvent.begin();
+        ctx.bind(localAddress, promise);
+        bindEvent.end();
+        if (bindEvent.shouldCommit()) {
+            bindEvent.setChannel(ctx.channel());
+            bindEvent.setLocalAddress(localAddress);
+            bindEvent.commit();
+        }
+    }
+
+    @Override
+    public void connect(
+            ChannelHandlerContext ctx,
+            SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
+        final ConnectEvent connectEvent = new ConnectEvent();
+        connectEvent.begin();
+        ctx.connect(remoteAddress, localAddress, promise);
+        connectEvent.end();
+        if (connectEvent.shouldCommit()) {
+            connectEvent.setChannel(ctx.channel());
+            connectEvent.setRemoteAddress(remoteAddress);
+            connectEvent.setLocalAddress(localAddress);
+            connectEvent.commit();
+        }
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) {
+        final DisconnectEvent disconnectEvent = new DisconnectEvent();
+        disconnectEvent.begin();
+        ctx.disconnect(promise);
+        disconnectEvent.end();
```

Comment:
so I was a wondering a bit what exactly this "event" should be used for... Should it tell us how long it took to do the actual bind ? If so we should call all the `end()` methods in a `ChannelFutureListener` that is attaches to the promise

Reply:
Problem with ending in a listener is that you get the wrong stacktrace, right @egahlin? but maybe that's ok.

Reply:
I guess the question is what this event should represent here.